preflight_check/check.py

```diff
@@ -0,0 +1,254 @@
+from crmsh import utils as crmshutils
+from crmsh import bootstrap as crmshboot
+
+
+def check(context):
+    """
+    Check environment and cluster state if related options enabled
+    """
+    if context.env_check:
+        check_environment()
+    if context.cluster_check:
+        check_cluster()
+    print()
+
+
+def check_environment():
+    """
+    A set of functions to check environment
+    """
+    print("\n============ Checking environment ============")
+    check_my_hostname_resolves()
+    check_time_service()
+    check_watchdog()
+    check_firewall()
+
+
+def check_my_hostname_resolves():
+    """
+    Check hostname resolvable
+    """
+    task_inst = task.TaskCheck("Checking hostname resolvable")
+    with task_inst.run():
+        if not crmshboot.my_hostname_resolves():
+            task_inst.error('''Hostname "{}" is unresolvable.
+  Please add an entry to /etc/hosts or configure DNS.'''.format(utils.this_node()))
+
+
+def check_time_service():
+    """
+    Check time service
+    """
+    task_inst = task.TaskCheck("Checking time service")
+    with task_inst.run():
+        timekeepers = ('chronyd.service', 'ntp.service', 'ntpd.service')
+        timekeeper = None
+        for tk in timekeepers:
+            if crmshutils.service_is_available(tk):
+                timekeeper = tk
+                break
+
+        if timekeeper is None:
+            task_inst.warn("No NTP service found.")
+        else:
+            task_inst.info("{} is available".format(timekeeper))
+            if crmshutils.service_is_enabled(timekeeper):
+                task_inst.info("{} is enabled".format(timekeeper))
+            else:
+                task_inst.warn("{} is disabled".format(timekeeper))
+
+            if crmshutils.service_is_active(timekeeper):
+                task_inst.info("{} is active".format(timekeeper))
+            else:
+                task_inst.warn("{} is not active".format(timekeeper))
+
+
+def check_watchdog():
+    """
+    Verify watchdog device
+    """
+    task_inst = task.TaskCheck("Checking watchdog")
+    with task_inst.run():
+        if not crmshutils.service_is_active("sbd.service") and not crmshboot.check_watchdog():
+            task_inst.warn("Watchdog device must be configured if want to use SBD!")
+
+
+def check_port_open(task, firewall_type):
+    """
+    Check whether corosync port is blocked by iptables
+    """
+    ports = utils.corosync_port_list()
+    if not ports:
+        task.error("Can not get corosync's port")
+        return
+
+    if firewall_type == "firewalld":
+        rc, out, err = crmshutils.get_stdout_stderr('firewall-cmd --list-port')
+        if rc != 0:
+            task.error(err)
+        for p in ports:
+            if re.search(' {}/udp'.format(p), out):
+                task.info("UDP port {} is opened in firewalld".format(p))
+            else:
+                task.error("UDP port {} should open in firewalld".format(p))
+    if firewall_type == "SuSEfirewall2":
```

Comment:
elif is better, or need return after condition "firewalld"

Reply:
Changed